src/golang/lib/workflow/operator/base.go

```diff
@@ -0,0 +1,304 @@
+	"github.com/aqueducthq/aqueduct/lib/collections/operator"
+	"github.com/aqueducthq/aqueduct/lib/collections/operator/function"
+	"github.com/aqueducthq/aqueduct/lib/collections/operator_result"
+	"github.com/aqueducthq/aqueduct/lib/collections/shared"
+	"github.com/aqueducthq/aqueduct/lib/database"
+	"github.com/aqueducthq/aqueduct/lib/job"
+	"github.com/aqueducthq/aqueduct/lib/vault"
+	"github.com/aqueducthq/aqueduct/lib/workflow/artifact"
+	"github.com/aqueducthq/aqueduct/lib/workflow/utils"
+	"github.com/dropbox/godropbox/errors"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+type baseOperator struct {
+	dbOperator *operator.DBOperator
+
+	// These fields are set to nil in the preview case.
+	resultWriter operator_result.Writer
+	resultID     uuid.UUID
+
+	metadataPath string
+	jobName      string
+
+	inputs              []artifact.Artifact
+	outputs             []artifact.Artifact
+	inputContentPaths   []string
+	inputMetadataPaths  []string
+	outputContentPaths  []string
+	outputMetadataPaths []string
+
+	jobManager    job.JobManager
+	vaultObject   vault.Vault
+	storageConfig *shared.StorageConfig
+	db            database.Database
+
+	resultsPersisted bool
+}
+
+func (bo *baseOperator) Type() operator.Type {
+	return bo.dbOperator.Spec.Type()
+}
+
+func (bo *baseOperator) Name() string {
+	return bo.dbOperator.Name
+}
+
+func (bo *baseOperator) ID() uuid.UUID {
+	return bo.dbOperator.Id
+}
+
+func (bo *baseOperator) Ready(ctx context.Context) bool {
+	for _, inputArtifact := range bo.inputs {
+		if !inputArtifact.Computed(ctx) {
+			return false
+		}
+	}
+	return true
+}
+
+// A catch-all for execution states that are the system's fault.
+// Logs an internal message so that we can debug.
+func unknownSystemFailureExecState(err error, logMsg string) *shared.ExecutionState {
+	log.Errorf("Execution had system failure: %s. %v", logMsg, err)
+
+	failureType := shared.SystemFailure
+	return &shared.ExecutionState{
+		Status:      shared.FailedExecutionStatus,
+		FailureType: &failureType,
+		Error: &shared.Error{
+			Context: fmt.Sprintf("%v", err),
+			Tip:     shared.TipUnknownInternalError,
+		},
+	}
+}
+
+// fetchExecState assumes that the operator has been computed already.
+func (bo *baseOperator) fetchExecState(ctx context.Context) *shared.ExecutionState {
+	var execState shared.ExecutionState
+	err := utils.ReadFromStorage(
+		ctx,
+		bo.storageConfig,
+		bo.metadataPath,
+		&execState,
+	)
+	if err != nil {
+		// Treat this as a system internal error since operator metadata was not found
+		return unknownSystemFailureExecState(
+			err,
+			"Unable to read operator metadata from storage. Operator may have failed before writing metadata.",
+		)
+	}
+	return &execState
+}
+
+// GetExecState takes a more wholelistic view of the operator's status than the job manager does,
+// and can be called at any time. Because of this, the logic for figuring out the correct state is
+// a little more involved.
```

Comment:
@likawind @saurav-c @hsubbaraj-spiral @cw75 I changed the logic here since the last review, when stamping out test failures. Can you double check the logic here?